File: ports/python/src/unicode_python/security/casing.py
"""UAX #21 case mapping (``toLower`` / ``toUpper``), mirroring
``Unicode.Casing``.

Full case mappings from ``SpecialCasing.txt`` (one-to-many and
context/locale-dependent rows) combined with the simple case mappings in
``UnicodeData.txt`` field 13 (lowercase) / 12 (uppercase). Context predicates
(Final_Sigma, After_Soft_Dotted, More_Above, Not_Before_Dot, After_I) use
canonical combining class plus the ``Cased`` and ``Soft_Dotted`` properties
from ``DerivedCoreProperties.txt``.

This is a shared primitive: ``bip39-canonical`` uses ``toLower(default)``, and
the ``locale-case-inversion`` / ``case-expansion-mismatch`` form detectors
build on ``toLower`` / ``toUpper``.
"""

from dataclasses import dataclass
from enum import Enum
from pathlib import Path

from .identity.ucd import ccc

_DATA_DIR = Path(__file__).resolve().parent.parent / "data"


def _read_data_file(name: str) -> str:
    with (_DATA_DIR / name).open("r", encoding="utf-8") as handle:
        return handle.read()


class Locale(Enum):
    """The locales SpecialCasing.txt distinguishes. ``DEFAULT`` covers
    everything not tagged Turkish / Azeri / Lithuanian."""

    DEFAULT = "default"
    TURKISH = "turkish"
    AZERI = "azeri"
    LITHUANIAN = "lithuanian"


_LOCALE_CONDITIONS = frozenset({"tr", "az", "lt"})


@dataclass(frozen=True, slots=True)
class _Row:
    code: int
    lower: tuple[int, ...]
    title: tuple[int, ...]
    upper: tuple[int, ...]
    conditions: tuple[str, ...]


# ─────────────────────────────────────────────────────────────────────
# Data tables (SpecialCasing rows, simple case mappings, Cased/Soft_Dotted)
# ─────────────────────────────────────────────────────────────────────


def _parse_special_casing() -> dict[int, list[_Row]]:
    rows: dict[int, list[_Row]] = {}
    for raw in _read_data_file("SpecialCasing.txt").split("\n"):
        line = raw.split("#", 1)[0].strip()
        if not line:
            continue
        fields = [f.strip() for f in line.split(";")]
        if len(fields) < 4:
            continue
        code = int(fields[0], 16)
        lower = tuple(int(tok, 16) for tok in fields[1].split())
        title = tuple(int(tok, 16) for tok in fields[2].split())
        upper = tuple(int(tok, 16) for tok in fields[3].split())
        conditions = tuple(fields[4].split()) if len(fields) > 4 and fields[4] else ()
        rows.setdefault(code, []).append(
            _Row(code=code, lower=lower, title=title, upper=upper, conditions=conditions)
        )
    return rows


def _parse_simple_case_mappings() -> tuple[dict[int, int], dict[int, int]]:
    """Return (lowercase, uppercase) maps from UnicodeData.txt fields 13/12.
    Codepoints absent from a map lowercase / uppercase to themselves."""
    lower: dict[int, int] = {}
    upper: dict[int, int] = {}
    for line in _read_data_file("UnicodeData.txt").split("\n"):
        if not line:
            continue
        fields = line.split(";")
        if len(fields) < 15:
            continue
        cp = int(fields[0], 16)
        if fields[12]:
            upper[cp] = int(fields[12], 16)
        if fields[13]:
            lower[cp] = int(fields[13], 16)
    return lower, upper


def _parse_derived_property(name: str) -> list[tuple[int, int]]:
    out: list[tuple[int, int]] = []
    for raw in _read_data_file("DerivedCoreProperties.txt").split("\n"):
        line = raw.split("#", 1)[0].strip()
        if not line:
            continue
        parts = line.split(";", 1)
        if len(parts) < 2 or parts[1].strip() != name:
            continue
        field = parts[0].strip()
        dots = field.find("..")
        if dots < 0:
            cp = int(field, 16)
            out.append((cp, cp))
        else:
            out.append((int(field[:dots], 16), int(field[dots + 2 :], 16)))
    out.sort(key=lambda r: r[0])
    return out


_SPECIAL_ROWS: dict[int, list[_Row]] | None = None
_SIMPLE_LOWER: dict[int, int] | None = None
_SIMPLE_UPPER: dict[int, int] | None = None
_CASED: list[tuple[int, int]] | None = None
_SOFT_DOTTED: list[tuple[int, int]] | None = None


def _special_rows() -> dict[int, list[_Row]]:
    global _SPECIAL_ROWS
    if _SPECIAL_ROWS is None:
        _SPECIAL_ROWS = _parse_special_casing()
    return _SPECIAL_ROWS


def _simple_maps() -> tuple[dict[int, int], dict[int, int]]:
    global _SIMPLE_LOWER, _SIMPLE_UPPER
    if _SIMPLE_LOWER is None or _SIMPLE_UPPER is None:
        _SIMPLE_LOWER, _SIMPLE_UPPER = _parse_simple_case_mappings()
    return _SIMPLE_LOWER, _SIMPLE_UPPER


def _in_ranges(ranges: list[tuple[int, int]], cp: int) -> bool:
    return any(lo <= cp <= hi for lo, hi in ranges)


def _is_cased(cp: int) -> bool:
    global _CASED
    if _CASED is None:
        _CASED = _parse_derived_property("Cased")
    return _in_ranges(_CASED, cp)


def _is_soft_dotted(cp: int) -> bool:
    global _SOFT_DOTTED
    if _SOFT_DOTTED is None:
        _SOFT_DOTTED = _parse_derived_property("Soft_Dotted")
    return _in_ranges(_SOFT_DOTTED, cp)


def simple_lowercase(cp: int) -> int:
    return _simple_maps()[0].get(cp, cp)


def simple_uppercase(cp: int) -> int:
    return _simple_maps()[1].get(cp, cp)


# ─────────────────────────────────────────────────────────────────────
# Context predicates (UAX #21). ``rev_prefix`` is the preceding codepoints
# nearest-first; ``suffix`` the strictly-following ones.
# ─────────────────────────────────────────────────────────────────────


def _more_above_after(suffix: list[int]) -> bool:
    for cp in suffix:
        c = ccc(cp)
        if c == 230:
            return True
        if c == 0:
            return False
    return False


def _after_soft_dotted(rev_prefix: list[int]) -> bool:
    for cp in rev_prefix:
        if _is_soft_dotted(cp):
            return True
        c = ccc(cp)
        if c in (0, 230):
            return False
    return False


def _after_i(rev_prefix: list[int]) -> bool:
    for cp in rev_prefix:
        if cp == 0x0049:
            return True
        c = ccc(cp)
        if c in (0, 230):
            return False
    return False


def _before_dot(suffix: list[int]) -> bool:
    for cp in suffix:
        if cp == 0x0307:
            return True
        if ccc(cp) == 0:
            return False
    return False


def _has_cased_before(rev_prefix: list[int]) -> bool:
    for cp in rev_prefix:
        if _is_cased(cp):
            return True
        if ccc(cp) == 0:
            return False
    return False


def _has_cased_after(suffix: list[int]) -> bool:
    for cp in suffix:
        if _is_cased(cp):
            return True
        if ccc(cp) == 0:
            return False
    return False


def _final_sigma(rev_prefix: list[int], suffix: list[int]) -> bool:
    return _has_cased_before(rev_prefix) and not _has_cased_after(suffix)


def _locale_matches(loc: Locale, conds: tuple[str, ...]) -> bool:
    if not any(c in _LOCALE_CONDITIONS for c in conds):
        return True
    return any(
        (c == "tr" and loc is Locale.TURKISH)
        or (c == "az" and loc is Locale.AZERI)
        or (c == "lt" and loc is Locale.LITHUANIAN)
        for c in conds
    )


def _conditions_hold(
    loc: Locale, rev_prefix: list[int], suffix: list[int], conds: tuple[str, ...]
) -> bool:
    if not _locale_matches(loc, conds):
        return False
    for c in conds:
        if c in _LOCALE_CONDITIONS:
            continue
        if c == "Final_Sigma":
            ok = _final_sigma(rev_prefix, suffix)
        elif c == "Not_Final_Sigma":
            ok = not _final_sigma(rev_prefix, suffix)
        elif c == "After_Soft_Dotted":
            ok = _after_soft_dotted(rev_prefix)
        elif c == "More_Above":
            ok = _more_above_after(suffix)
        elif c == "Not_Before_Dot":
            ok = not _before_dot(suffix)
        elif c == "After_I":
            ok = _after_i(rev_prefix)
        else:  # unrecognised context token — never matches
            ok = False
        if not ok:
            return False
    return True


def _find_special_row(
    loc: Locale, rev_prefix: list[int], suffix: list[int], cp: int
) -> _Row | None:
    candidates = _special_rows().get(cp)
    if candidates is None:
        return None
    # UAX #21: a conditional row whose conditions hold outranks the
    # unconditional row for the same codepoint.
    for row in candidates:
        if row.conditions and _conditions_hold(loc, rev_prefix, suffix, row.conditions):
            return row
    for row in candidates:
        if not row.conditions:
            return row
    return None


def _lower_codepoint(
    loc: Locale, rev_prefix: list[int], suffix: list[int], cp: int
) -> list[int]:
    row = _find_special_row(loc, rev_prefix, suffix, cp)
    if row is not None:
        return list(row.lower)
    return [simple_lowercase(cp)]


def _upper_codepoint(
    loc: Locale, rev_prefix: list[int], suffix: list[int], cp: int
) -> list[int]:
    row = _find_special_row(loc, rev_prefix, suffix, cp)
    if row is not None:
        return list(row.upper)
    return [simple_uppercase(cp)]


def to_lower(loc: Locale, cps: list[int]) -> list[int]:
    """Lowercase a codepoint sequence under ``loc`` (UAX #21 full mapping)."""
    out: list[int] = []
    rev_prefix: list[int] = []
    for index, cp in enumerate(cps):
        suffix = cps[index + 1 :]
        out.extend(_lower_codepoint(loc, rev_prefix, suffix, cp))
        rev_prefix.insert(0, cp)
    return out


def to_upper(loc: Locale, cps: list[int]) -> list[int]:
    """Uppercase a codepoint sequence under ``loc`` (UAX #21 full mapping)."""
    out: list[int] = []
    rev_prefix: list[int] = []
    for index, cp in enumerate(cps):
        suffix = cps[index + 1 :]
        out.extend(_upper_codepoint(loc, rev_prefix, suffix, cp))
        rev_prefix.insert(0, cp)
    return out

File: ports/lua/src/unicode_lua/security/casing.lua
-- UAX #21 case mapping (toLower / toUpper), mirroring `Unicode.Casing`.
--
-- Full case mappings from SpecialCasing.txt (one-to-many and context/locale-
-- dependent rows) combined with the simple case mappings in UnicodeData.txt
-- field 13 (lowercase) / 12 (uppercase).  Context predicates (Final_Sigma,
-- After_Soft_Dotted, More_Above, Not_Before_Dot, After_I) use canonical
-- combining class plus the Cased and Soft_Dotted properties from
-- DerivedCoreProperties.txt.
--
-- Shared primitive: bip39-canonical uses toLower(default); the
-- locale-case-inversion form detector builds on lower_codepoint.

local datapath = require("unicode_lua.datapath")
local ucd = require("unicode_lua.security.identity.ucd")

local ccc = ucd.ccc

local M = {}

-- The locales SpecialCasing.txt distinguishes.  DEFAULT covers everything not
-- tagged Turkish / Azeri / Lithuanian.
local Locale = {
  DEFAULT = "default",
  TURKISH = "turkish",
  AZERI = "azeri",
  LITHUANIAN = "lithuanian",
}
M.Locale = Locale

local LOCALE_CONDITIONS = { tr = true, az = true, lt = true }

-- ─────────────────────────────────────────────────────────────────────
-- Parsing helpers
-- ─────────────────────────────────────────────────────────────────────

local function iter_lines(text)
  return (text .. "\n"):gmatch("([^\n]*)\n")
end

local function trim(s)
  return (s:gsub("^%s*(.-)%s*$", "%1"))
end

local function split(s, sep)
  local out = {}
  local start = 1
  while true do
    local i = s:find(sep, start, true)
    if i == nil then
      out[#out + 1] = s:sub(start)
      break
    end
    out[#out + 1] = s:sub(start, i - 1)
    start = i + #sep
  end
  return out
end

local function parse_hex_tokens(s)
  local out = {}
  for tok in s:gmatch("%S+") do
    out[#out + 1] = tonumber(tok, 16)
  end
  return out
end

-- ─────────────────────────────────────────────────────────────────────
-- Data tables
-- ─────────────────────────────────────────────────────────────────────

local _special_rows = nil
local _simple_lower = nil
local _simple_upper = nil
local _cased = nil
local _soft_dotted = nil

local function parse_special_casing()
  local rows = {}
  for raw in iter_lines(datapath.read("SpecialCasing.txt")) do
    local line = trim(split(raw, "#")[1])
    if line ~= "" then
      local fields = split(line, ";")
      if #fields >= 4 then
        local code = tonumber(trim(fields[1]), 16)
        local lower = parse_hex_tokens(fields[2])
        local title = parse_hex_tokens(fields[3])
        local upper = parse_hex_tokens(fields[4])
        local conditions = {}
        if #fields > 4 and trim(fields[5]) ~= "" then
          for tok in trim(fields[5]):gmatch("%S+") do
            conditions[#conditions + 1] = tok
          end
        end
        if rows[code] == nil then
          rows[code] = {}
        end
        local list = rows[code]
        list[#list + 1] = { code = code, lower = lower, title = title, upper = upper, conditions = conditions }
      end
    end
  end
  return rows
end

local function parse_simple_case_mappings()
  local lower = {}
  local upper = {}
  for line in iter_lines(datapath.read("UnicodeData.txt")) do
    if line ~= "" then
      local fields = split(line, ";")
      if #fields >= 15 then
        local cp = tonumber(fields[1], 16)
        if fields[13] ~= "" then
          upper[cp] = tonumber(fields[13], 16)
        end
        if fields[14] ~= "" then
          lower[cp] = tonumber(fields[14], 16)
        end
      end
    end
  end
  return lower, upper
end

local function parse_derived_property(name)
  local out = {}
  for raw in iter_lines(datapath.read("DerivedCoreProperties.txt")) do
    local line = trim(split(raw, "#")[1])
    if line ~= "" then
      local parts = split(line, ";")
      if #parts >= 2 and trim(parts[2]) == name then
        local field = trim(parts[1])
        local dots = field:find("..", 1, true)
        if dots == nil then
          local cp = tonumber(field, 16)
          out[#out + 1] = { cp, cp }
        else
          out[#out + 1] = { tonumber(field:sub(1, dots - 1), 16), tonumber(field:sub(dots + 2), 16) }
        end
      end
    end
  end
  table.sort(out, function(a, b) return a[1] < b[1] end)
  return out
end

local function special_rows()
  if _special_rows == nil then
    _special_rows = parse_special_casing()
  end
  return _special_rows
end

local function simple_maps()
  if _simple_lower == nil then
    _simple_lower, _simple_upper = parse_simple_case_mappings()
  end
  return _simple_lower, _simple_upper
end

local function in_ranges(ranges, cp)
  for _, r in ipairs(ranges) do
    if r[1] <= cp and cp <= r[2] then
      return true
    end
  end
  return false
end

local function is_cased(cp)
  if _cased == nil then
    _cased = parse_derived_property("Cased")
  end
  return in_ranges(_cased, cp)
end

local function is_soft_dotted(cp)
  if _soft_dotted == nil then
    _soft_dotted = parse_derived_property("Soft_Dotted")
  end
  return in_ranges(_soft_dotted, cp)
end

function M.simple_lowercase(cp)
  local lower, _ = simple_maps()
  return lower[cp] or cp
end

function M.simple_uppercase(cp)
  local _, upper = simple_maps()
  return upper[cp] or cp
end

-- ─────────────────────────────────────────────────────────────────────
-- Context predicates (UAX #21).  `rev_prefix` is preceding codepoints
-- nearest-first; `suffix` is the strictly-following ones in order.
-- ─────────────────────────────────────────────────────────────────────

local function more_above_after(suffix)
  for _, cp in ipairs(suffix) do
    local c = ccc(cp)
    if c == 230 then
      return true
    end
    if c == 0 then
      return false
    end
  end
  return false
end

local function after_soft_dotted(rev_prefix)
  for _, cp in ipairs(rev_prefix) do
    if is_soft_dotted(cp) then
      return true
    end
    local c = ccc(cp)
    if c == 0 or c == 230 then
      return false
    end
  end
  return false
end

local function after_i(rev_prefix)
  for _, cp in ipairs(rev_prefix) do
    if cp == 0x0049 then
      return true
    end
    local c = ccc(cp)
    if c == 0 or c == 230 then
      return false
    end
  end
  return false
end

local function before_dot(suffix)
  for _, cp in ipairs(suffix) do
    if cp == 0x0307 then
      return true
    end
    if ccc(cp) == 0 then
      return false
    end
  end
  return false
end

local function has_cased_before(rev_prefix)
  for _, cp in ipairs(rev_prefix) do
    if is_cased(cp) then
      return true
    end
    if ccc(cp) == 0 then
      return false
    end
  end
  return false
end

local function has_cased_after(suffix)
  for _, cp in ipairs(suffix) do
    if is_cased(cp) then
      return true
    end
    if ccc(cp) == 0 then
      return false
    end
  end
  return false
end

local function final_sigma(rev_prefix, suffix)
  return has_cased_before(rev_prefix) and not has_cased_after(suffix)
end

local function has_locale_condition(conds)
  for _, c in ipairs(conds) do
    if LOCALE_CONDITIONS[c] then
      return true
    end
  end
  return false
end

local function locale_matches(loc, conds)
  if not has_locale_condition(conds) then
    return true
  end
  for _, c in ipairs(conds) do
    if (c == "tr" and loc == Locale.TURKISH)
      or (c == "az" and loc == Locale.AZERI)
      or (c == "lt" and loc == Locale.LITHUANIAN) then
      return true
    end
  end
  return false
end

local function conditions_hold(loc, rev_prefix, suffix, conds)
  if not locale_matches(loc, conds) then
    return false
  end
  for _, c in ipairs(conds) do
    if not LOCALE_CONDITIONS[c] then
      local ok
      if c == "Final_Sigma" then
        ok = final_sigma(rev_prefix, suffix)
      elseif c == "Not_Final_Sigma" then
        ok = not final_sigma(rev_prefix, suffix)
      elseif c == "After_Soft_Dotted" then
        ok = after_soft_dotted(rev_prefix)
      elseif c == "More_Above" then
        ok = more_above_after(suffix)
      elseif c == "Not_Before_Dot" then
        ok = not before_dot(suffix)
      elseif c == "After_I" then
        ok = after_i(rev_prefix)
      else
        -- Unrecognised context token — never matches.
        ok = false
      end
      if not ok then
        return false
      end
    end
  end
  return true
end

local function find_special_row(loc, rev_prefix, suffix, cp)
  local candidates = special_rows()[cp]
  if candidates == nil then
    return nil
  end
  -- A conditional row whose conditions hold outranks the unconditional row.
  for _, row in ipairs(candidates) do
    if #row.conditions > 0 and conditions_hold(loc, rev_prefix, suffix, row.conditions) then
      return row
    end
  end
  for _, row in ipairs(candidates) do
    if #row.conditions == 0 then
      return row
    end
  end
  return nil
end

local function copy_list(t)
  local out = {}
  for _, v in ipairs(t) do
    out[#out + 1] = v
  end
  return out
end

function M.lower_codepoint(loc, rev_prefix, suffix, cp)
  local row = find_special_row(loc, rev_prefix, suffix, cp)
  if row ~= nil then
    return copy_list(row.lower)
  end
  return { M.simple_lowercase(cp) }
end

function M.upper_codepoint(loc, rev_prefix, suffix, cp)
  local row = find_special_row(loc, rev_prefix, suffix, cp)
  if row ~= nil then
    return copy_list(row.upper)
  end
  return { M.simple_uppercase(cp) }
end

-- Slice cps[from..#cps] into a new 1-indexed table.
local function suffix_from(cps, from)
  local out = {}
  for i = from, #cps do
    out[#out + 1] = cps[i]
  end
  return out
end

function M.to_lower(loc, cps)
  local out = {}
  local rev_prefix = {}
  for index = 1, #cps do
    local cp = cps[index]
    local suffix = suffix_from(cps, index + 1)
    for _, r in ipairs(M.lower_codepoint(loc, rev_prefix, suffix, cp)) do
      out[#out + 1] = r
    end
    table.insert(rev_prefix, 1, cp)
  end
  return out
end

function M.to_upper(loc, cps)
  local out = {}
  local rev_prefix = {}
  for index = 1, #cps do
    local cp = cps[index]
    local suffix = suffix_from(cps, index + 1)
    for _, r in ipairs(M.upper_codepoint(loc, rev_prefix, suffix, cp)) do
      out[#out + 1] = r
    end
    table.insert(rev_prefix, 1, cp)
  end
  return out
end

return M
